### PR TITLE
Add the section to encode unordered setters as a shape

### PR DESCRIPTION
**Pull request by Atry**
In addition to #174, I also added a section to encode unordered setters as a shape.

### CODE REVIEW

**Comment by fredemmott** [comment]
- I think we need to settle on 'long-lived'/observability/'resource' terminology first
- This seems to contradict https://github.com/hhvm/hsl/pull/174

**Comment by fredemmott** [comment]
> * I think we need to settle on 'long-lived'/observability/'resource' terminology first
> * This seems to contradict [Add the section to encode setters as a vec of setter interface #174](https://github.com/hhvm/hsl/pull/174)

Depending on how the first point is resolved, the second *might* disappear.

**Comment by fredemmott** [comment]
The contradiction I see is that: these should be shapes... unless there's a vec of them?

**Comment by Atry** [comment]
I updated the PR. Instead of defining the terminology long-lived, I added a section to define short-lived, because the default encoding for handles and pointers would be closeable classes, while short-lived is a special case that requires a special definition

**Comment by fredemmott** [comment]
> I think we need to settle on 'long-lived'/observability/'resource' terminology first

While the definition of short-lived helps, I still strongly feel that the terms 'long-lived' and 'short-lived' are misleading here. They imply that "how long does the object exist?" is the key factor, and it isn't relevant. The definition clears up that this isn't the case, but it's still counter to their normal meaning and makes the document hard to read.

Observability still seems like the right way to define the separation. If the user can't tell if the side effects have happened or not, it doesn't matter that they happened - it's an implementation detail. What problems are you seeing with observability?

Assuming there are solid problems with observability: we need clearer terms - ideally clear enough to remove the need for a clarifying definition. Would using 'with side-effects' and 'side-effect-free' throughout be workable?

**Review by fredemmott — CHANGES_REQUESTED** [review]
Per previous comment

**Comment by Atry** [comment]
In terms of observability, shall we use the terminology "Reference Transparency"? The terminology describes exactly whether side effects are observable.

**Comment by fredemmott** [comment]
>  "Referential Transparency"?

I'm not really able to think things through today and heading off sick - that said, my initial reaction is that this seems correct - though perhaps 'value-like' or '[POD](https://en.cppreference.com/w/cpp/types/is_pod)-like' might be more approachable, if less precise.

If you think referential transparency is a better choice, it's probably worth adding a small note explaining it, and that it is purely based on the C side of this, given all Hack objects are always passed by reference

**Comment by Atry** [comment]
I prefer "referential transparency" to "side-effect-free" because functions accepting a referential transparent type do perform side effects, for example

1. A setter performs the side effect to modify the content of a referential transparent type
2. `posix_spawn` accepts `posix_spawn_file_actions_t` and performs the side effect to create a process

`1.` is not a violation of referential transparency because it's a local side effect in the expression to create the referential transparent type. `2.` is not a violation because `posix_spawn` does not distinguish `posix_spawn_file_actions_t` pointers whose addresses are different while the pointed contents are the same.

**Comment by Atry** [comment]
> If you think referential transparency is a better choice, it's probably worth adding a small note explaining it, and that it is purely based on the C side of this, given all Hack objects are always passed by reference

I added a paragraph to explain the meaning of "referential transparent" in C and Hack respectively.

File: src/os/README.md [hunk 1]
```diff
@@ -112,7 +111,8 @@ inspiration for several decisions in this library.
 
 | C type | Hack type |
 |---|---|
-| Pointers to referential transparent C `struct`s | `vec` |
+| Pointers to referential transparent C `struct`s whose setters are unordered | `shape` |
+| Other pointers to referential transparent C `struct`s | `vec` |
```
Comment (fredemmott) [inline src/os/README.md:115]: this wording is assuming that the majority of structs we care about:
- have setter functions as apposed to just containing properties that we can directly manipulate
- have order that matters beyond the C ABI - e.g. the order of properties in a `sockaddr_un` are ordered, this matters to the C ABI, but it doesn't matter for Hack at all.

First point I think needs revision; second is primarily a readability issue.

File: src/os/README.md [hunk 2]
```diff
@@ -189,3 +189,38 @@ final class posix_spawn_file_actions_adddup2 implements PosixSpawnFileActionsSet
 
 type posix_spawn_file_actions_t = vec<PosixSpawnFileActionsSetter>;
 ```
+
+#### Special encoding for unordered setters as a `shape`
+
+For some C data types the setters are orthogonal to each other, and the order to
```
Comment (fredemmott) [inline src/os/README.md:195]: This should swap with the above, or entirely remove 'default' wording for either case.

**Review by fredemmott — CHANGES_REQUESTED** [review]

File: src/os/README.md (hunk 2 above)
Comment (Atry) [reply src/os/README.md]: Order swapped

File: src/os/README.md (hunk 1 above)
Comment (Atry) [reply src/os/README.md]: Added "`struct`s of primitive types or other `struct`s"

**Pull request closed without merging**